aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -1,53 +1,81 @@
-#include "TensorIterator.h"
+#include <ATen/native/TensorIterator.h>
 
+#include <array>
 #include <ATen/ExpandUtils.h>
 #include <ATen/Parallel.h>
 
 namespace at {
 
-void TensorIterator::reorder_dimensions() {
-  // Sort the dimensions based on the sum-of-strides in ascending order. NOTE:
-  // that this inverts the order of C-contiguous tensors. strides[0] is the
-  // fastest moving dimension instead of strides[ndim - 1].
+struct DimCounter {
+  DimCounter(IntList shape, Range range);
 
-  auto sum_of_strides = SmallVector<double, 6>(ndim(), 0.0);
-  for (int dim = 0; dim < ndim(); dim++) {
-    double sum = 0.0;
-    for (const auto& op : operands_) {
-      if (op.stride_bytes.size() == 0) continue;
-      sum += op.stride_bytes[dim];
-    }
+  void increment(const std::array<int64_t, 2>& step);
+  bool is_done() const;
+  std::array<int64_t, 2> max_step() const;
 
-    // Weight each dimension by its index. Given two dimensions with equal
-    // some of strides, this preserves the given relative ordering.
-    sum += (ndim() - dim - 1) / (double)ndim();
+  IntList shape;
+  Range range;
+  DimVector values;
+  int64_t offset;
+};
 
-    sum_of_strides[dim] = sum;
-  }
+using DimMask = TensorIterator::DimMask;
+using PtrVector = TensorIterator::PtrVector;
+using loop_t = TensorIterator::loop_t;
+using loop2d_t = TensorIterator::loop2d_t;
 
-  // initialize perm with 0, 1, 2, ...
-  perm_.resize(ndim());
-  std::iota(std::begin(perm_), std::end(perm_), 0);
-
-  std::sort(std::begin(perm_), std::end(perm_), [&](size_t i1, size_t i2) {
-    return sum_of_strides[i1] < sum_of_strides[i2];
-  });
+void TensorIterator::reorder_dimensions() {
+  // Sort the dimensions based on strides in ascending order with reduced dims
```

Comment:
It looks like you changed the strategy for permuting the dimensions. Does it affect runtime for pointwise operations?

What if the first argument is the single one that has strides different than the (possibly many) other arguments?

Reply:
Previously we sorted based on the sum-of-strides. I changed it to only swap dimensions that make the ordering no-worse across all arguments (and better for at least one argument), with stride 0 not counting and reduced dimensions moved to the front. This more closely matches the stride sorting behavior of THC and NumPy, with the exception of the reduced dimensions behavior.

I don't expect this to affect the runtime for pointwise operations. Most orderings remain the same, and contiguous dims get moved to the front, which is the most important aspect.

I'm not sure this is the final strategy for ordering strides, but this logic makes it easier to support reductions.